misc/build-web-index

```diff
@@ -17,33 +17,45 @@
 # detect when a file lacks such a title and fail loudly, instead of just
 # omitting that file from the index.
 
-set -e
+import re
+import sys
 
-# The non-quotation is intentional: I am trying to get bash to split on words so
-# that `grep` takes the arguments as separate even though they were passed as
-# one thing to this script.
-docs=$1
-examples=$2
-libraries=$3
+def file_block(files):
+  for fname in files:
+    if fname.startswith("doc/"):
+      link_name = fname[len("doc/"):]
+    else:
+      link_name = fname
+    with open(fname, 'r') as f:
+      found_title = False
+      for line in f.readlines():
+        title = re.match(r"' *# ?(.*)", line)
+        if title:
+          print(f"- [{fname}]({link_name}.html) {title.group(1)}")
+          found_title = True
+          break
+      if not found_title:
+        raise ValueError(f"File {fname} did not have a title (top-level Markdown heading)")
```

Comment:
The `found_title` + `if` can actually be replaced by an [`else` block to the `for` loop](https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops).